File: lib/src/main/java/com/sd/lib/tag_view/FTagView.java
package com.sd.lib.tag_view;

import android.content.Context;
import android.util.AttributeSet;
import android.widget.FrameLayout;

import java.lang.reflect.Modifier;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public class FTagView extends FrameLayout implements ITagView
{
    private final String mViewTag;
    private final ItemManager mItemManager;

    public FTagView(Context context, AttributeSet attrs)
    {
        super(context, attrs);
        mViewTag = Utils.getObjectId(this);
        mItemManager = new InternalItemManager();
    }

    @Override
    public final String getViewTag()
    {
        return mViewTag;
    }

    @Override
    public final ItemManager getItemManager()
    {
        return mItemManager;
    }

    protected <T extends Item> T createItem(Class<T> clazz)
    {
        try
        {
            return clazz.newInstance();
        } catch (Exception e)
        {
            throw new RuntimeException("Create instance failed for class:" + clazz.getName(), e);
        }
    }

    private final class InternalItemManager implements ItemManager
    {
        private final Map<Class<? extends Item>, Item> nItemHolder = new ConcurrentHashMap<>();

        @Override
        public synchronized <T extends Item> T findItem(Class<T> clazz)
        {
            checkItemClass(clazz);

            final Item item = nItemHolder.get(clazz);
            if (item == null)
                return null;
            return (T) item;
        }

        @Override
        public synchronized <T extends Item> T getItem(Class<T> clazz)
        {
            checkItemClass(clazz);

            Item item = nItemHolder.get(clazz);
            if (item == null)
            {
                item = createItem(clazz);
                if (item == null)
                    throw new NullPointerException("createItem method return null");

                nItemHolder.put(clazz, item);
                item.init(FTagView.this);
            }
            return (T) item;
        }

        @Override
        public synchronized void clearItem()
        {
            for (Item item : nItemHolder.values())
            {
                item.destroy();
            }
            nItemHolder.clear();
        }
    }

    private static <T extends Item> void checkItemClass(Class<T> clazz)
    {
        if (clazz == null)
            throw new IllegalArgumentException("clazz is null");

        if (clazz.isInterface())
            throw new IllegalArgumentException("clazz is interface " + clazz);

        if (Modifier.isAbstract(clazz.getModifiers()))
            throw new IllegalArgumentException("clazz is abstract " + clazz);
    }
}

File: lib/src/main/java/com/sd/lib/tag_view/FTagView.kt
package com.sd.lib.tag_view

import android.content.Context
import android.util.AttributeSet
import android.widget.FrameLayout
import java.lang.reflect.Modifier
import java.util.*
import java.util.concurrent.ConcurrentHashMap

class FTagView : FrameLayout, ITagView {
    /** 保存[Item] */
    private val _itemHolder: MutableMap<Class<out Item>, Item> = ConcurrentHashMap()

    constructor(context: Context, attrs: AttributeSet?) : super(context, attrs)

    override val viewTag: String by lazy {
        Utils.getObjectId(this@FTagView)
    }

    /**
     * 查找指定类型的Item，如果不存在则返回null
     */
    fun <T : Item> findItem(clazz: Class<T>): T? {
        checkItemClass(clazz)
        val item = _itemHolder[clazz] ?: null
        return item as T
    }

    /**
     * 返回指定类型的Item，如果不存在则创建对象返回
     */
    fun <T : Item> getItem(clazz: Class<T>): T {
        checkItemClass(clazz)
        val cacheItem = _itemHolder[clazz]
        if (cacheItem != null) {
            return cacheItem as T
        }

        synchronized(this@FTagView) {
            val item = createItem(clazz)
            _itemHolder[clazz] = item
            item.initItem(this@FTagView)
            return item
        }
    }

    /**
     * 移除并销毁Item
     */
    fun <T : Item> removeItem(clazz: Class<T>): T? {
        checkItemClass(clazz)
        synchronized(this@FTagView) {
            val cacheItem = _itemHolder.remove(clazz) ?: return null
            cacheItem.destroyItem()
            return cacheItem as T
        }
    }

    /**
     * 清空并销毁所有Item
     */
    fun destroyItem() {
        synchronized(this@FTagView) {
            for (item in _itemHolder.values) {
                item.destroyItem()
            }
            _itemHolder.clear()
        }
    }

    /**
     * Item必须有空参数的构造方法
     */
    interface Item {
        /**
         * 初始化
         */
        fun initItem(tagView: ITagView)

        /**
         * 销毁
         */
        fun destroyItem()
    }

    companion object {
        private fun <T : Item> checkItemClass(clazz: Class<T>) {
            require(!clazz.isInterface) { "clazz is interface $clazz" }
            require(!Modifier.isAbstract(clazz.modifiers)) { "clazz is abstract $clazz" }
        }

        private fun <T : Item> createItem(clazz: Class<T>): T {
            return try {
                clazz.newInstance()
            } catch (e: Exception) {
                throw RuntimeException("Create instance failed for class:" + clazz.name, e)
            }
        }
    }
}